proposals/testing/NNNN-polling-confirmations.md

```diff
@@ -0,0 +1,512 @@
+///     `duration` must be greater than 0.
+///   - interval: The minimum amount of time to wait between polling attempts.
+///     If nil, this uses whatever value is specified under the last
+///     ``PollingUntilFirstPassConfigurationTrait`` or
+///     ``PollingUntilStopsPassingConfigurationTrait`` added to the test or
+///     suite.
+///     If no such trait has been added, then polling will wait at least
+///     1 millisecond between polling attempts.
+///     `interval` must be greater than 0.
+///   - isolation: The actor to which `body` is isolated, if any.
+///   - sourceLocation: The source location to whych any recorded issues should
+///     be attributed.
+///   - body: The function to invoke.
+///
+/// - Throws: A `PollingFailedError` if the `body` does not return true within
+///   the polling duration.
+///
+/// - Returns: The last non-nil value returned by `body`.
+///
+/// Use polling confirmations to check that an event while a test is running in
+/// complex scenarios where other forms of confirmation are insufficient. For
+/// example, waiting on some state to change that cannot be easily confirmed
+/// through other forms of `confirmation`.
+@available(_clockAPI, *)
+@discardableResult
+public func confirmation<R>(
+  _ comment: Comment? = nil,
+  until stopCondition: PollingStopCondition,
+  within duration: Duration? = nil,
+  pollingEvery interval: Duration? = nil,
+  isolation: isolated (any Actor)? = #isolation,
+  sourceLocation: SourceLocation = #_sourceLocation,
+  _ body: @escaping () async throws -> sending R?
+) async throws -> R
+```
+
+### New `PollingStopCondition` enum
+
+A new enum type, `PollingStopCondition` will be defined, specifying when to stop
+polling before the duration has elapsed. Additionally, if the early stop
+condition isn't fulfilled before the duration elapses, then this also defines
+how the confirmation should be handled.
+
+```swift
+/// A type defining when to stop polling early.
+/// This also determines what happens if the duration elapses during polling.
+public enum PollingStopCondition: Sendable {
+  /// Evaluates the expression until the first time it returns true.
+  /// If it does not pass once by the time the timeout is reached, then a
+  /// failure will be reported.
+  case firstPass
+
+  /// Evaluates the expression until the first time it returns false.
+  /// If the expression returns false, then a failure will be reported.
+  /// If the expression only returns true before the timeout is reached, then
+  /// no failure will be reported.
+  /// If the expression does not finish evaluating before the timeout is
+  /// reached, then a failure will be reported.
+  case stopsPassing
+}
+```
+
+### New Error Type
+
+A new error type, `PollingFailedError` to be thrown when the polling
+confirmation doesn't pass:
+
+```swift
+/// A type describing an error thrown when polling fails.
+public struct PollingFailedError: Error, Sendable, CustomIssueRepresentable {}
```

Comment:
`CustomIssueRepresentable` is not a `public` protocol so it doesn't need to be shown here and I'd omit it to avoid confusion.